[WEBSITE-688] - Document the Plugin permission management in a new page (#2678)
jenkins-infra/jenkins.io: content/doc/developer/plugin-governance/managing-permissions.adoc

---
layout: developersection
title: Managing plugin permissions
references:
- url: https://github.com/jenkins-infra/repository-permissions-updater
  title: Repository Permissions Updater
- url: ../adopt-a-plugin
  title: Adopting a plugin
---

This page documents managing permissions for plugin maintainers.
There are 2 main permissions available to plugin contributors:

> * GitHub permissions. 

Reviewer: Make it clear that this is a definition term followed by the definition of the term.  I wish that the layout of ASCIIDoc definition lists ("labeled lists") were better, but this is technique provides an easier to read layout that adoc labeled lists (at least for me)
```suggestion
* **GitHub permissions**. 
```

  It includes merging pull requests, push to the repositories, managing GitHub applications, etc.
  The effective list of permissions depends on the link:https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization[permission levels], see below.

> * Release Permissions which are needed to deploy releases to the link:https://repo.jenkins-ci.org/webapp/#/home[Jenkins artifact repository].

Reviewer: ```suggestion
* **Release Permissions** which are needed to deploy releases to the link:https://repo.jenkins-ci.org/webapp/#/home[Jenkins artifact repository].
```

  Snapshots can be deployed by contributors without this permission.

NOTE: If you want to adopt a plugin, please see link:../adopt-a-plugin[Adopt a Plugin] instead.

== GitHub Permissions

GitHub permissions in Jenkins are managed by GitHub teams.
Almost every plugin has a `${plugin-id} Developers` team created specifically for this plugin.
This team may have a different link:https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization[permission level] depending on time of creation.
For all new repositories we tend to grant `Admin` permissions at the moment, and other plugin teams can get permissions elevated upon a request.

How to request GitHub permissions?
You will need to send a permission transfer request and to specify the permission recipient's GitHub account and Jenkins LDAP(aka Jenkins Jira) account there.

* Request permissions change in Jenkins Jira
** Project: `INFRA`, Component: `github`

> * Request permissions in the Jenkins developer mailing list. 

Reviewer: ```suggestion
* Request permissions in the linkhttps://groups.google.com/forum/#!forum/jenkinsci-dev[:Jenkins developer mailing list].
```

Permission change requests should be ideally done by one of the active plugin maintainers.
Otherwise, an explicit approval will be required before the permission transfer happens.
If there are multiple active maintainers in a plugin, Jenkins GitHub administrators will usually wait for a while to collect approvals from a number of them.

=== Valid approvals by maintainers

> The following ways of approval are generally recognized as valid ones:

Reviewer: ```suggestion
The following approvals are generally recognized as valid:
```

* Confirmation in a Jenkins Jira ticket (permission request or other ticket)
* Explicit approval in a GitHub pull request or issue
** The reply should be sent from a GitHub account associated with a maintainer's Jenkins LDAP account

> * Confirmation in the developer mailing list.

Reviewer: ```suggestion
* Confirmation in the link:https://groups.google.com/forum/#!forum/jenkinsci-dev[developer mailing list].
```

** The reply should be sent from the email listed in the Jenkins LDAP account

Other types of approval (emails from different addresses, email forwards by requesters, etc.) will be reviewed and verified by `jenkinsci` GitHub administrators.

== Release Permissions

Permissions to upload plugin releases are independent of GitHub push access and maintained in the link:https://github.com/jenkins-infra/repository-permissions-updater[Repository Permissions Updater] repository.

To request upload permissions for a new maintainer:

. If you have never done it before, 
  you log in at least in once with your Jenkins account into the link:https://repo.jenkins-ci.org/webapp/#/home[Jenkins artifact repository].
  Any modification to the permission files will be ineffective until then.

> . File a PR on that repository in link:https://github.com/jenkins-infra/repository-permissions-updater[Repository Permissions Updater] as described in the README of that repository. Without this, you'll be unable to release your plugin.

Reviewer: The references to 'that' were not completely clear to me.  I think that I've understood them.  You're welcome to ignore this comment if I've misunderstood them.
```suggestion
. File a PR in the link:https://github.com/jenkins-infra/repository-permissions-updater[Repository Permissions Updater] for the specific plugin repository which needs the permission change.
  Refer to the link:https://github.com/jenkins-infra/repository-permissions-updater/blob/master/README.md[Repository Permissions Updater README] for more detailed instructions. 
  Once the permissions are updated, you'll be able to release your plugin.
```

//TODO(oleg-nenashev): Add CoC violations and so on?

By default we *DO NOT* remove upload permissions for current or past maintainers.

> It can be done based on the explicit approval of the previous maintainer or based on the discussion in the Developer mailing list.

Reviewer: ```suggestion
It can be done based on the explicit approval of the previous maintainer or based on a discussion in the link:https://groups.google.com/forum/#!forum/jenkinsci-dev[Developer mailing list].
```

---
layout: developersection
title: Managing plugin permissions
references:
- url: https://github.com/jenkins-infra/repository-permissions-updater
  title: Repository Permissions Updater
- url: ../adopt-a-plugin
  title: Adopting a plugin
---

This page documents managing permissions for plugin maintainers.
There are 2 main permissions available to plugin contributors:

* **GitHub permissions**. 
  It includes merging pull requests, push to the repositories, managing GitHub applications, etc.
  The effective list of permissions depends on the link:https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization[permission levels], see below.

> * **Release Permissions** which are needed to deploy releases to the link:https://repo.jenkins-ci.org/webapp/#/home[Jenkins artifact repository].

Reviewer 2: Do you need to mention this is how plugins are released to the update center as well?

Reviewer 3: Probably better in https://jenkins.io/doc/developer/publishing/artifact-repository/

== GitHub Permissions

GitHub permissions in Jenkins are managed by GitHub teams.
Almost every plugin has a `<pluginId> Developers` team created specifically for this plugin.
This team may have a different link:https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization[permission level] depending on time of creation.
For all new repositories we tend to grant `Admin` permissions at the moment, and other plugin teams can get permissions elevated upon a request.

How to request GitHub permissions?
You will need to send a permission transfer request and to specify the permission recipient's GitHub account and Jenkins LDAP(aka Jenkins Jira) account there.

* Request permissions change in Jenkins Jira
** Project: `INFRA`, Component: `github`
* Request permissions in the link:https://groups.google.com/forum/#!forum/jenkinsci-dev[:Jenkins developer mailing list].

Permission change requests should be ideally done by one of the active plugin maintainers.
Otherwise, an explicit approval will be required before the permission transfer happens.

> If there are multiple active maintainers in a plugin, Jenkins GitHub administrators will usually wait for a while to collect approvals from a number of them.

Reviewer 3: I think this section should explicitly distinguish between:

* An existing maintainer wants to add co-maintainers, committers, or transfer ownership to another user (perhaps because the other user asked them for it directly, but doesn't matter)
* Someone wants to take over an abandoned plugin / unresponsive maintainer

The former is a simple admin task, so it's good enough to have a PR to RPU, or a Jira issue filed by one of the RPU uploader accounts (or both). There is never a reason to go through the dev list for this.

The latter needs dev list involvement per adopt-this-plugin, etc. as it needs more visibility and awareness, as there is a greater potential for it to go wrong.

Reviewer: I'd be fine with this change as is, or also fine with the additional information proposed by @daniel-beck .  I think this is a positive step whether or not it includes the recommended addition.

== Release Permissions

Permissions to upload plugin releases are independent of GitHub push access and maintained in the link:https://github.com/jenkins-infra/repository-permissions-updater[Repository Permissions Updater] repository.

To request upload permissions for a new maintainer:

. If you have never done it before, 
  you log in at least in once with your Jenkins account into the link:https://repo.jenkins-ci.org/webapp/#/home[Jenkins artifact repository].
  Any modification to the permission files will be ineffective until then.
. File a PR in the link:https://github.com/jenkins-infra/repository-permissions-updater[Repository Permissions Updater] for the specific plugin repository which needs the permission change.
  Refer to the link:https://github.com/jenkins-infra/repository-permissions-updater/blob/master/README.md[Repository Permissions Updater README] for more detailed instructions. 
  Once the permissions are updated, you'll be able to release your plugin.

//TODO(oleg-nenashev): Add CoC violations and so on?

By default we *DO NOT* remove upload permissions for current or past maintainers.

> It can be done based on the explicit approval of the previous maintainer or based on a discussion in the link:https://groups.google.com/forum/#!forum/jenkinsci-dev[Developer mailing list].

Reviewer 4: is this needed? i.e. I've maintained a plugin for over a year, there's ~5 other people with commit / release access who haven't touched it in that time.
Shouldn't I just be able to remove them?

Author: Should or should not is a quite separate question. I am just trying to interpret the current governance model. @rodrigc has already created a JEP with a proposal to change it, and I am in favor of changes (team/company ownership, maintainer removal and so on). But it is rather a further change after the Governance board is reestablished

Reviewer: I interpret this to be a description of default behavior in the project, not mandatory behavior by everyone that participates in the project.  Since you are a "previous maintainer" and a "current maintainer", I think you could announce to the dev list that you intend to remove specific people from the list, then proceed with the removal.  If they object, they can be added again.

Reviewer 3: At least while I was involved in RPU, this wasn't that strict; maintainers could remove other maintainers by filing a PR and waiting a week or so (or getting explicit confirmation). I wanted to wait a bit to prevent some weird dispute from escalating. Obviously it's not good enough in the "vacation" case, but then there's a track of the previous maintainership status and could be reinstated. Obviously at that point it needs escalating/mediation anyway.

* **Release Permissions** which are needed to deploy releases to the link:/doc/developer/publishing/artifact-repository/[Jenkins artifact repository].
  Snapshots can be deployed by contributors without this permission.

NOTE: If you want to adopt a plugin, please see link:../adopt-a-plugin[Adopt a Plugin] instead.

== GitHub Permissions

GitHub permissions in Jenkins are managed by GitHub teams.
Almost every plugin has a `<pluginId> Developers` team created specifically for this plugin.
This team may have a different link:https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization[permission level] depending on time of creation.
For all new repositories we tend to grant `Admin` permissions at the moment, and other plugin teams can get permissions elevated upon a request.

=== Changing GitHub permissions as a maintainer

Any active maintainer is eligible to add more contributors to the GitHub repository.
There are multiple ways to do that:

> . Via GitHub Web Interface, if you have admin permissions in the `${plugin-id} Developers` team

Author: ```suggestion
. Via GitHub Web Interface, if you have admin permissions in the `<pluginId> Developers` team
```
